From db3d1aff4e963ac19d965681b022e1f09a00a51a Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Tue, 20 Feb 2024 10:42:42 +0530
Subject: [PATCH] wifi: mac80211: add net stats in ethtool stats

Add net tx stats(tx_net_packets) to ethtool stats.
This tx stats accounts the packets that comes from net layer to mac80211.

Also add net rx stats(rx_net_packets) to ethtool stats.
This rx stats accounts the packets passed to net layer from mac80211 and ath12k driver.

sample log:
root@OpenWrt:/# ethtool -S wlan2
NIC statistics:
     rx_packets: 19
     rx_bytes: 1635
     rx_duplicates: 0
     rx_fragments: 7
     rx_dropped: 0
     tx_packets: 6
     tx_bytes: 908
     tx_filtered: 0
     tx_retry_failed: 0
     tx_retries: 0
     sta_state: 0
     txrate: 0
     rxrate: 0
     signal: 0
     channel: 0
     noise: 18446744073709551615
     ch_time: 18446744073709551615
     ch_time_busy: 18446744073709551615
     ch_time_ext_busy: 18446744073709551615
     ch_time_rx: 18446744073709551615
     ch_time_tx: 18446744073709551615
     tx_dev_dropped: 8
     rx_dev_dropped: 0
     tx_net_packets: 32
     rx_net_packets: 9
     tx_errors: 0
     rx_errors: 0
Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 include/net/mac80211.h |  2 ++
 net/mac80211/ethtool.c | 11 +++++++++++
 2 files changed, 13 insertions(+)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 9e6f65a..6be6868 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -7860,6 +7860,8 @@ enum ieee80211_gstrings_stats {
 	IEEE80211_CH_TIME_TX,
 	IEEE80211_TX_DEV_DROPPED,
 	IEEE80211_RX_DEV_DROPPED,
+	IEEE80211_TX_NET_PACKETS,
+	IEEE80211_RX_NET_PACKETS,
 	IEEE80211_STATS_MAX,
 };
 
diff --git a/net/mac80211/ethtool.c b/net/mac80211/ethtool.c
index bb583a3..0d20ff2 100644
--- a/net/mac80211/ethtool.c
+++ b/net/mac80211/ethtool.c
@@ -71,6 +71,8 @@ static const char ieee80211_gstrings_sta_stats[][ETH_GSTRING_LEN] = {
 	[IEEE80211_CH_TIME_TX] = "ch_time_tx",
 	[IEEE80211_TX_DEV_DROPPED] = "tx_dev_dropped",
 	[IEEE80211_RX_DEV_DROPPED] = "rx_dev_dropped",
+	[IEEE80211_TX_NET_PACKETS] = "tx_net_packets",
+	[IEEE80211_RX_NET_PACKETS] = "rx_net_packets",
 };
 
 #define STA_STATS_LEN	ARRAY_SIZE(ieee80211_gstrings_sta_stats)
@@ -103,6 +105,10 @@ static void ieee80211_get_stats(struct net_device *dev,
 	struct survey_info survey;
 	int q;
 	unsigned long int i;
+	struct rtnl_link_stats64 rtnl_stats = {0};
+
+	dev_fetch_sw_netstats(&rtnl_stats, dev->tstats);
+
 #define STA_STATS_SURVEY_LEN 7
 
 	memset(data, 0, sizeof(u64) * STA_STATS_LEN);
@@ -231,6 +237,11 @@ do_survey:
 	data[IEEE80211_RX_DEV_DROPPED] = sdata->rx_dropped;
 	i++;
 
+	data[IEEE80211_TX_NET_PACKETS] = rtnl_stats.tx_packets;
+	i++;
+	data[IEEE80211_RX_NET_PACKETS] = rtnl_stats.rx_packets;
+	i++;
+
 	if (WARN_ON(i != STA_STATS_LEN))
 		return;
 
-- 
2.34.1

